### PR TITLE
qemu_img_lock_reject_boot: check output without using exceptions

### PR DESCRIPTION
**Pull request by lkotek**
Method `create` from `qemu_vm.py` tries to connect to QMP monitor right after initial cheking of QEMU process to be alive. This can bring timing issue as sometimes QEMU process dies slowly (observed in case of luks image) after this initial check when connections to QMP monitors are being established. As a result `MonitorConnectError` exception can be triggered instead of `VMCreateError` anticipated by `qemu_img_lock_reject_boot` test case. 

This PR changes the logic of test case, so all exceptions are caught and expected error messages are checked differently.

**Context**

- This PR replaces https://github.com/avocado-framework/avocado-vt/pull/3616 and solves the issue on tp-qemu level. 
- Please see relevant discussion here: https://github.com/avocado-framework/avocado-vt/pull/3616#discussion_r1099765273.

ID: 2156320

Signed-off-by: Lukas Kotek <lkotek@redhat.com>

### CODE REVIEW

**Comment by lkotek** [comment]
Hi @luckyh, we continued discussing this issue and the conclusion was to put the output checking into `except` block. It resolves the problem, because `MonitorConnectError` is still triggered (as we observed it in case of luks image usage), so we only need to catch *any* exception and then use the logic @YongxueHong recommended.

**Comment by yanan-fu** [comment]
@qingwangrh Could you help review this patch ?  And help verify it with all disk format. Thanks!

**Review by yanan-fu — APPROVED** [review]
Ack

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Comment by lkotek** [comment]
@yanan-fu @YongxueHong Thank you! But please do not merge yet. I found another problem – at the moment `vm2.process.get_output().strip()` is called we get `NoneType` as vm2 is destroyed at the moment. 

I am looking for the way how to workaround it.

**Comment by YongxueHong** [comment]
> @yanan-fu @YongxueHong Thank you! But please do not merge yet. I found another problem – at the moment `vm2.process.get_output().strip()` is called we get `NoneType` as vm2 is destroyed at the moment.
> 
> I am looking for the way how to workaround it.

Hi @lkotek 
Yeah,  you're right, I also found it yesterday.
The process will be closed in the process of `_cleanup` and the job `close` from `class Spawn` in the 
`aexpect` lib will remove the all the related files which include the output file.
```
    def close(self, sig=signal.SIGKILL):
        """
        Kill the child process if it's alive and remove temporary files.

        :param sig: The signal to send the process when attempting to kill it.
        """
        if not self.closed:
            self.kill(sig=sig)
            # Wait for the server to exit
            wait_for_lock(self.lock_server_running_filename)
            # Call all cleanup routines
            for hook in self.close_hooks:
                hook(self)
            # Close reader file descriptors
            self._close_reader_fds()
            self.reader_fds = {}
            # Remove all used files
            if 'AEXPECT_DEBUG' not in os.environ:
                shutil.rmtree(os.path.join(BASE_DIR, f'aexpect_{self.a_id}'))
            self._close_aexpect_helper()
            self.closed = True
```
Then the get_output would get a `None` type as return value.
```
    def get_output(self):
        """
        Return the STDOUT and STDERR output of the process so far.
        """
        try:
            with open(self.output_filename, 'rb') as output_file:
                return output_file.read().decode(self.encoding,
                                                 'backslashreplace')
        except IOError:
            return None
```
Thanks for pointing it.

**Comment by YongxueHong** [comment]
Hi @lkotek 
I will change the request for this PR to avoid other maintainers merging it.
Thanks.

**Review by YongxueHong — CHANGES_REQUESTED** [review]
Hi @lkotek
I will change the request for this PR to avoid other maintainers merging it.
Thanks.

File: qemu/tests/qemu_img_lock_reject_boot.py
Comment (qingwangrh) [inline qemu/tests/qemu_img_lock_reject_boot.py:66]: Catch all exceptions is accepted, but i wonder why the exception message comes from vm.process? does it cover all kinds of VM creation failed?

Comment (lkotek) [reply qemu/tests/qemu_img_lock_reject_boot.py:66]: Yep, there is a corner case of `MonitorConnectError` when the VM is actually already dead, but it also means we can't get the expected error message as it is outside of context of `MonitorConnectError` excetion raised.

**Comment by lkotek** [comment]
@YongxueHong @yanan-fu Thank you all for help with debugging this issue. I just successfully tested the solution using the avocado's `debug.log` for checking the error message with following result:
```
JOB ID     : 5c04d20c23c2b41fea31c3bdb262390d21b56033
JOB LOG    : /root/avocado/job-results/job-2023-03-01T03.21-5c04d20/job.log
 (1/1) Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35: STARTED
 (1/1) Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35: PASS (144.59 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2023-03-01T03.21-5c04d20/results.html
```
Key part in the log (log level is set to debug, so it is a bit 'messy'):
```
[stdlog] 2023-03-01 03:23:21,035 avocado.test INFO | Verify qemu-img write lock err msg.
[stdlog] 2023-03-01 03:23:21,035 avocado.utils.process INFO | Running 'cat /root/avocado/job-results/latest/test-results/*Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35/debug.log | grep '\[qemu output\]' | grep -v 'warning''
[stdlog] 2023-03-01 03:23:21,042 avocado.utils.process DEBUG| [stdout] [stdlog] 2023-03-01 03:21:21,482 avocado.virttest.qemu_vm INFO | [qemu output] qemu-kvm: -device {"driver": "scsi-hd", "id": "image1", "drive": "drive_image1", "write-cache": "on"}: Failed to get "write" lock
[stdlog] 2023-03-01 03:23:21,043 avocado.utils.process INFO | Command 'cat /root/avocado/job-results/latest/test-results/*Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35/debug.log | grep '\[qemu output\]' | grep -v 'warning'' finished with 0 after 0.003905928s
[stdlog] 2023-03-01 03:23:21,043 avocado.utils.process DEBUG| [stdout] [stdlog] 2023-03-01 03:21:21,482 avocado.virttest.qemu_vm INFO | [qemu output] Is another process using the image [/home/kvm_autotest_root/images/rhel900-64-virtio-scsi.luks]?
```

**Comment by lkotek** [comment]
Last change is due to `inspekt` complained about escaping sequence, it is fixed now and also tested again:
```
JOB LOG    : /root/avocado/job-results/job-2023-03-01T04.06-c78c1c7/job.log
 (1/1) Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35: STARTED
 (1/1) Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35: PASS (144.83 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

File: qemu/tests/qemu_img_lock_reject_boot.py
Comment (YongxueHong) [inline qemu/tests/qemu_img_lock_reject_boot.py:37]: You could get the debug.log through `test.logfile` https://github.com/avocado-framework/avocado/blob/eb5418015a304f517822bdeab9570c31919302a0/avocado/core/test.py#L393

Comment (lkotek) [reply qemu/tests/qemu_img_lock_reject_boot.py:37]: Let me check this right now. I noticed this in the docs before, but I thought it contains only the actual name of logfile, not the full path. But comment here clearly refers about path. Just a moment :wink:

Comment (lkotek) [reply qemu/tests/qemu_img_lock_reject_boot.py]: @YongxueHong It works :slightly_smiling_face: Thanks for the catch, it helps to simplify things.

```
JOB LOG    : /root/avocado/job-results/job-2023-03-01T05.06-5788fa2/job.log
 (1/1) Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35: STARTED
 (1/1) Host_RHEL.m9.u0.v0.luks.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.qemu_img_lock.reject_boot_same_img_twice.q35: PASS (144.82 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

**Review by YongxueHong — APPROVED** [review]

**Review by yanan-fu — APPROVED** [review]
LGTM

**Comment by yanan-fu** [comment]
Thanks all!

**Pull request merged**
